File: baekjoon/9000/9506.cpp
#include <bits/stdc++.h>
#define loop(i,n) for(int i=0;i<n;++i)
using namespace std;

void perfect(int n)
{
    int sum = 0;
    vector<int> vt;
    for (int i = 1; i <= n / 2;i++) {
        if (n % i == 0) {
            sum += i;
            vt.push_back(i);
        }
    }
    if (sum == n) {
        cout << n << " = ";
        loop(i, vt.size()) {
            cout << vt[i];
            if (i != vt.size() - 1) cout << " + ";
            else cout << '\n';
        }
    }
    else
        cout << n << " is NOT perfect." << '\n';
}
int main()
{
    int n;
    while (1) {
        cin >> n;
        if (n == -1) break;
        perfect(n);
    }
    return 0;
}

File: baekjoon/9000/9506.rs
use std::io::*;

struct Scanner {
    it: std::str::SplitAsciiWhitespace<'static>,
}

impl Scanner {
    fn new() -> Self {
        let mut s = String::new();
        stdin().read_to_string(&mut s).ok();
        Self { it: s.leak().split_ascii_whitespace() }
    }
    fn read<T: std::str::FromStr>(&mut self) -> T {
        self.it.next().unwrap().parse::<T>().ok().unwrap()
    }
}

fn main() {
    let mut sc = Scanner::new();
    let mut bw = BufWriter::new(stdout().lock());
    loop {
        let n = sc.read::<i32>();
        if n == -1 {
            break;
        }
        let mut div = (1..=n / 2).filter(|&x| n % x == 0).collect::<Vec<_>>();
        if n == div.iter().sum() {
            write!(bw, "{n} = ").ok();
            let last = div.pop().unwrap();
            for d in div {
                write!(bw, "{d} + ").ok();
            }
            writeln!(bw, "{last}").ok();
        } else {
            writeln!(bw, "{n} is NOT perfect.").ok();
        }
    }
}
